internal/controller/servermaintenanceset_controller.go

```diff
@@ -0,0 +1,247 @@
+		// Check if the server already got a maintenance
+		for _, maintenance := range maintenancelist.Items {
+			if maintenance.Spec.ServerRef.Name == server.Name {
+				log.V(1).Info("Server is already in maintenance", "server", server.Name)
+				continue out
+			}
+		}
+		maintenance := metalv1alpha1.ServerMaintenance{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%d", replica.Name, i),
+				Namespace: replica.Namespace,
+			},
+		}
+		opResult, err := controllerutil.CreateOrPatch(ctx, r.Client, &maintenance, func() error {
+			metautils.SetLabels(&maintenance, map[string]string{ServerMaintenanceSetFinalizer: replica.Name})
+			maintenance.Spec = metalv1alpha1.ServerMaintenanceSpec{
+				ServerRef:                       &v1.LocalObjectReference{Name: server.Name},
+				Policy:                          replica.Spec.Template.Policy,
+				ServerPower:                     replica.Spec.Template.ServerPower,
+				ServerBootConfigurationTemplate: replica.Spec.Template.ServerBootConfigurationTemplate,
+			}
+			return controllerutil.SetControllerReference(replica, &maintenance, r.Scheme, controllerutil.WithBlockOwnerDeletion(false))
+		})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to create or patch serverMaintenance %s: %w", maintenance.Name, err))
+			continue
+		}
+		log.V(1).Info("Created or patched ServerMaintenance", "ServerMaintenance", maintenance.Name, "Operation", opResult)
+	}
+	if len(errs) > 0 {
+		return ctrl.Result{}, fmt.Errorf("errors occurred during servermaintenances creation: %v", errs)
+	}
+
+	// Fetch the list of maintenances again to get the updated status
+	maintenancelist, err = r.getOwnedMaintenances(ctx, replica)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Fetched ServerMaintenances", "Count", len(maintenancelist.Items))
+	if modified, err := r.patchStatus(ctx, replica, calculateStatus(maintenancelist.Items)); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Reconciled ServerMaintenanceSet", "Name", replica.Name)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceSetReconciler) patchStatus(ctx context.Context, replica *metalv1alpha1.ServerMaintenanceSet, status metalv1alpha1.ServerMaintenanceSetStatus) (bool, error) {
+	if replica.Status == status {
+		return false, nil
+	}
+	base := replica.DeepCopy()
+	replica.Status = status
+	if err := r.Status().Patch(ctx, replica, client.MergeFrom(base)); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *ServerMaintenanceSetReconciler) getOwnedMaintenances(ctx context.Context, replica *metalv1alpha1.ServerMaintenanceSet) (*metalv1alpha1.ServerMaintenanceList, error) {
+	maintenancelist := &metalv1alpha1.ServerMaintenanceList{}
+	opts := []client.ListOption{
+		client.InNamespace(replica.Namespace),
+		client.MatchingLabels{ServerMaintenanceSetFinalizer: replica.Name},
+	}
+	if err := r.List(ctx, maintenancelist, opts...); err != nil {
+		return nil, err
+	}
```

Comment:
You could use here https://github.com/ironcore-dev/controller-utils/blob/main/clientutils/clientutils.go#L441 to get all dependant objects.